docs: Add docs structure recommendations, update style guide

qmonnet opened the pull request:
Expand the docs for contributing to Cilium's documentation, by adding new items to the style guide (including common pitfalls to avoid), and by adding a new section with recommendations on the structure of the docs.

Most of the new additions were written by Sarah, I mostly took care of porting them to Cilium's documentation.

qmonnet committed:
docs: Add docs structure recommendations, update style guide

Expand the docs for contributing to Cilium's documentation, by adding
new items to the style guide (including common pitfalls to avoid), and
by adding a new section with recommendations on the structure of the
docs.

Most of the new additions were written by Sarah, I mostly took care of
porting them to Cilium's documentation.

Co-authored-by: Sarah Corleissen <sarah.corleissen@isovalent.com>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

bimmlerd commented on Documentation/contributing/docs/docsstructure.rst:
```diff
@@ -0,0 +1,98 @@
+a page expands it to include multiple concepts or tasks, then add new pages for
+individual concepts and tasks.
+
+If you're moving a page and changing its URL, make sure you update every link
```
Should this also mention the need to potentially update the link map from newer version to older versions?

qmonnet replied:
Do you mean the redirects?

Contributors don't usually have access to the relevant config (I think it's in the ReadTheDocs interface), but we could suggest asking on Slack if necessary

bimmlerd replied:
Yeah that's what I mean. That's a good point; yeah maybe asking on slack is the best option.

qmonnet replied:
OK, I added it. Thanks!

bimmlerd reviewed (approved):
one question, but non-blocking. This is awesome!

learnitall commented on Documentation/contributing/docs/docsstyle.rst:
```diff
@@ -253,6 +269,31 @@ Lists
 
+- Be consistent with periods at the end of list items. In general, omit periods
+  from bulleted list items unless the items are complete sentences. But if one
+  list item requires a period, use periods for all items
```
```suggestion
  list item requires a period, use periods for all items.
```

Does this need a period here?

qmonnet replied:
It does, thanks.

learnitall commented on Documentation/contributing/docs/docsstyle.rst:
```diff
@@ -253,6 +269,31 @@ Lists
+
+Use callouts effectively. For example, use the ``.. note::`` directive to
+highlight information that helps users in a specific context. Do not use it to
+avoid refactoring a section or paragraph.
```
What would be an example of a user using a note to avoid a refactor?

qmonnet replied:
I'll come up with one.

learnitall commented on Documentation/contributing/docs/docsstyle.rst:
```diff
@@ -270,3 +311,167 @@ Roles
 
     See `this GitHub issue <https://github.com/cilium/cilium/issues/1234>`__.
+
+Common pitfalls
```
```suggestion
Common Pitfalls
```

zacharysarah replied:
I don't love that we're putting subheadings in caps...but in the spirit of eating our own dogfood, at least it's consistent with the rest of the page. 😬

qmonnet replied:
Good catch, thank you.

> I don't love that we're putting subheadings in caps

Then let's change it? I don't have a strong preference, I picked one option to have one in the guide but that's about it. And it's not as if the docs were consistent in that regard anyway.

learnitall commented on Documentation/contributing/docs/docsstyle.rst:
```diff
@@ -270,3 +311,167 @@ Roles
+.. _Kubernetes style guide: https://kubernetes.io/docs/contribute/style/style-guide/
+.. _content best practices: https://kubernetes.io/docs/contribute/style/style-guide/#content-best-practices
+
+Specific language
```
What's the difference between this section and the sections "Use specific instructions rather than vague instructions" and "Use specific wording rather than vague wording"? I feel like those two sections are examples that could be put under this section.

zacharysarah replied:
@learnitall That's a good idea.

qmonnet replied:
True, I'll move them under this section. Thanks

learnitall reviewed (changes requested):
This is sweet! I have two nits and a couple of questions on the structure.

zacharysarah commented on Documentation/contributing/docs/docsstyle.rst:
```diff
@@ -20,9 +20,25 @@ documentation. They have several objectives:
+"color" instead of ":spelling:ignore:`colour`".
+
+Maintain a consistent style with the rest of the documentation when possible,
+or at least with the rest of the updated page, in any case.
```
```suggestion
or at least with the rest of the updated page.
```

zacharysarah commented on Documentation/contributing/docs/docsstyle.rst:
```diff
@@ -270,3 +311,167 @@ Roles
+Common pitfalls
+---------------
+
+There are best practices for writing documentation: try to follow them. In
```
Do or do not; there is no try. 
```suggestion
Follow best practices for writing documentation. In
```

qmonnet replied:
:+1: May the Docs be with you

zacharysarah commented on Documentation/contributing/docs/docsstyle.rst:
```diff
@@ -270,3 +311,167 @@ Roles
+There are best practices for writing documentation: try to follow them. In
+general, default to the `Kubernetes style guide`_, especially for `content best
+practices`_. Give particular attention to the recommendations from the
+following subsections.
```
```suggestion
practices`_. The follow subsections cover the most common feedback given for Cilium documentation PRs.
```

zacharysarah commented on Documentation/contributing/docs/docsstyle.rst:
```diff
@@ -270,3 +311,167 @@ Roles
+.. code-block:: rst
+
+  Feature A requires all pods to be painted blue. Consequently, the Agent must
+  apply its "paint" action to all pods. To make the Agent paing all pods blue,
```
typo
```suggestion
  apply its "paint" action to all pods. To make the Agent paint all pods blue,
```

zacharysarah reviewed (changes requested):
@qmonnet Thanks for doing this! Let's add #26904 as a followup once this merges.

qmonnet committed:
docs: Add subsection on capitalisation to style guide

We should have clear directives regarding capitalisation in the docs.
For example, we should follow the Kubernetes style guide on how to
capitalise API objects. Let's add a relevant subsection to the style
guide.

Examples from Sarah and Nick Young.

Fixes: #26904
Suggested-by: Sarah Corleissen <sarah.corleissen@isovalent.com>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

learnitall reviewed (approved):
Awesome looks great!

zacharysarah reviewed (approved):
Thanks for adding this, @qmonnet! 🔆
